File: src/claudebox_frontend/src/api/containers.js
/** Containers API client - workspace-scoped endpoints only. */

import {
  FETCH_TIMEOUT_CONTAINER_LIFECYCLE_MS,
  FETCH_TIMEOUT_INTERACTIVE_MS,
} from '../config/timing'
import { workspaceFetch } from './apiClient'

/** Fetch one container's registry entry; null when it is not registered. */
export async function getContainer(containerId, { signal } = {}) {
  const res = await workspaceFetch(`/containers/${containerId}`, {
    signal,
    timeoutMs: FETCH_TIMEOUT_INTERACTIVE_MS,
  })

  return res.ok ? res.json() : null
}

/** Delete (stop and remove) a container in the current workspace. */
export async function deleteContainer(containerId) {
  const res = await workspaceFetch(`/containers/${containerId}`, {
    method: 'DELETE',
    timeoutMs: FETCH_TIMEOUT_CONTAINER_LIFECYCLE_MS,
  })
  if (!res.ok) {
    throw new Error('Failed to delete container')
  }
  return res.json()
}

/** List containers registered to the current workspace. */
export async function listContainers() {
  const res = await workspaceFetch('/containers')
  if (!res.ok) {
    throw new Error(`Failed to list containers: ${res.status}`)
  }
  return res.json()
}

File: src/claudebox_daemon/handlers/containers.py
"""Container signal/proxy routes (workspace-scoped only)."""

from fastapi import APIRouter, Request

from ._models import StopContainerRequest
from ._shared import WorkspaceDep


workspace_router = APIRouter(prefix="/api/workspaces/{workspace_id}")


# Workspace-scoped routes
# ----------------------------------------------------------------------------------------------


@workspace_router.get("/containers")
async def list_containers(svc: WorkspaceDep) -> dict:
    """List every registered container for this workspace."""

    return {"containers": svc.container_service.list_all()}


@workspace_router.get("/containers/{container_id}")
async def get_container(svc: WorkspaceDep, container_id: str):
    """Return a single registered container for a workspace."""

    return svc.container_service.get(container_id)


@workspace_router.post("/containers/{container_id}/stop")
async def stop_container(svc: WorkspaceDep, container_id: str, body: StopContainerRequest):
    """Send SIGTERM with grace; container ends STOPPED, stays in the registry."""

    await svc.container_service.stop_container(container_id, **body.model_dump())

    return {"id": container_id, "status": "stopped"}


@workspace_router.post("/containers/{container_id}/kill")
async def kill_container(svc: WorkspaceDep, container_id: str):
    """Send SIGKILL; container ends STOPPED, stays in the registry."""

    await svc.container_service.kill_container(container_id)

    return {"id": container_id, "status": "stopped"}


@workspace_router.delete("/containers/{container_id}")
async def delete_container(svc: WorkspaceDep, container_id: str):
    """Gracefully stop and remove a container within a workspace."""

    await svc.container_service.stop_container(container_id)
    await svc.container_service.remove(container_id)

    return {"id": container_id, "status": "deleted"}


@workspace_router.api_route(
    "/containers/{container_id}/{path:path}",
    methods=["GET", "POST", "PUT", "PATCH", "DELETE"],
)
async def proxy(svc: WorkspaceDep, container_id: str, path: str, request: Request):
    """Reverse-proxy requests to a container backend within a workspace."""

    return await svc.container_service.forward(request, container_id, path)
